ui/src/hooks.ts

```diff
@@ -1,5 +1,36 @@
 import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
+import { useParams } from "react-router-dom";
 import { AppDispatch, RootState } from "store";
 
 export const useAppDispatch = () => useDispatch<AppDispatch>();
 export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
+
+export const useUnderlay = () => {
+  const { underlayName } = useParams<{ underlayName: string }>();
+  return useAppSelector((state) =>
+    state.underlays.find((underlay) => underlay.name === underlayName)
+  );
+};
+
+export const useUnderlayOrFail = () => {
```

Comment:
Could you add some comments here? What is the guidance on when to use this vs `useUnderlay`? Are there any considerations around the time during which the underlay is still loading?

Reply:
Done. The idea is that the UI enforces the existence of a valid underlay before allowing access to lower level pages so the useXOrFail versions allow the code to ignore the possibility of them being undefined.